Kernel/Graphics/QXLGraphicsAdapter.cpp

```diff
@@ -0,0 +1,463 @@
+    public:
+        MemSlot(QXLDevice&, u32, PhysicalAddress, size_t);
+    private:
+        u64 m_high_bits { 0 };
+        u8 m_generation { 0 };
+    };
+
+    void reset();
+
+    QXLGraphicsAdapter& m_adapter;
+    TypedMapping<volatile QXLRom> m_rom;
+    TypedMapping<volatile QXLRamHeader> m_ram_header;
+    PhysicalAddress m_vram_addr;
+    PhysicalAddress m_surface_addr;
+    size_t m_surface_size { 0 };
+    u16 m_io_base { 0 };
+    OwnPtr<Ring> m_command_ring;
+    OwnPtr<Ring> m_cursor_ring;
+    OwnPtr<Ring> m_release_ring;
+    OwnPtr<MemSlot> m_vram_slot;
+    OwnPtr<MemSlot> m_surface_slot;
+};
+
+UNMAP_AFTER_INIT bool QXLDevice::initialize(PCI::Address const& pci_address, RefPtr<FramebufferDevice>& framebuffer_device, RefPtr<Graphics::FramebufferConsole>& console)
+{
+    dbgln("Initialize qxl device {}.{}", pci_address.device(), pci_address.function());
+    auto rom_addr = PhysicalAddress(PCI::get_BAR2(pci_address) & 0xfffffffc);
+    auto rom_size = PCI::get_BAR_space_size(pci_address, 2);
+    m_vram_addr = PhysicalAddress(PCI::get_BAR0(pci_address) & 0xfffffffc);
+    m_rom = map_typed<volatile QXLRom>(rom_addr);
+    if (m_rom->magic != 0x4f525851)
+        return false;
+    m_ram_header = map_typed_writable<volatile QXLRamHeader>(m_vram_addr.offset(m_rom->ram_header_offset));
+    if (m_ram_header->magic != 0x41525851)
+        return false;
+    m_io_base = PCI::get_BAR3(pci_address) & 0xfff0;
+
+    bool surface_64_bits = false;
+    if (PCI::get_BAR_space_size(pci_address, 4) > 0) {
+        m_surface_addr = PhysicalAddress(PCI::get_BAR4(pci_address) & 0xfffffffc);
+        m_surface_size = PCI::get_BAR_space_size(pci_address, 4);
+        surface_64_bits = true;
+    } else {
+        m_surface_addr = PhysicalAddress(PCI::get_BAR1(pci_address) & 0xfffffffc);
+        m_surface_size = PCI::get_BAR_space_size(pci_address, 1);
+    }
+
+    dbgln("QXL: Device version {}.{}", (u32)m_rom->id, (u32)m_rom->update_id);
+    dbgln("QXL: Compression level: {}", (u32)m_rom->compression_level);
+    dbgln("QXL: Log level: {}", (u32)m_rom->log_level);
+    dbgln("QXL: ROM at {:} with {} bytes, VRAM at {:} with {} bytes", rom_addr, rom_size, m_vram_addr, (u32)m_rom->surface0_area_size);
+    dbgln("QXL: IO Base: 0x{:x}", m_io_base);
+    dbgln("QXL: Surface: {} with {} bytes ({} bits)", m_surface_addr, m_surface_size, surface_64_bits ? "64" : "32");
+
+    // This is a little weird, but we *must* map the framebuffer, otherwise setting up the memslot later will fail!
+    framebuffer_device = FramebufferDevice::create(m_adapter, 0, framebuffer_address(), 1024, 768, 1024 * sizeof(u32));
+    framebuffer_device->initialize();
+    console = Graphics::FramebufferConsole::initialize(framebuffer_address(), 1024, 768, 1024 * sizeof(u32));
+
+    m_command_ring = adopt_own_if_nonnull(new Ring(m_ram_header->cmd_ring_hdr, sizeof(m_ram_header->cmd_ring) / sizeof(m_ram_header->cmd_ring[0]), QXLIOCommand::NOTIFY_CMD, false));
```

Comment:
This should be changed to `new (nothrow) Ring`, otherwise the kernel will panic on allocation failure.

More preferably, you can also use `try_make<T>()` for creating `OwnPtr<T>` and `try_create<T>()` for creating `RefPtr<T>` if `T`'s constructor is public.

See a391175 for why it's needed